parsers/CY.py

```diff
@@ -12,64 +12,102 @@
+        time_value = arrow.get(time_str).replace(tzinfo='Asia/Nicosia').datetime
+
+        prods_value = {
+            'oil': prods_list[3],
+            'solar': prods_list[1],
+            'wind': prods_list[0],
+            'biomass': 0.0
+        }
+
+        # Because solar is explicitly listed as "Solar PV" (so no thermal with energy storage) and there
+        # is zero sunlight in the middle of the night (https://www.timeanddate.com/sun/cyprus/nicosia),
+        # we use the biomass+solar generation reported at 0:00 to determine the portion of biomass+solar
+        # which constitutes biomass
+        if len(data) == 0:
+            prods_value['biomass'] = prods_value['solar']
+            prods_value['solar'] = 0.0
+        else:
+            nocturnal_biomass_generation = data[0]['production']['biomass']
+            prods_value['solar'] -= nocturnal_biomass_generation
+            prods_value['biomass'] += nocturnal_biomass_generation
+        if prods_value['solar'] < 0.0:
+            # if there is (next to) no sunlight and biomass is lower than at midnight
+            prods_value['solar'] = 0.0
+
+        data.append({
+            'zoneKey': 'CY',
+            'production': prods_value,
+            'storage': {},
+            'source': 'tsoc.org.cy',
+            'datetime': time_value
+        })
+
+
+    def parse_html(self, html: str) -> list:
+        global EXPECTED_COLUMNS, COLUMNS_REGEX, TIMES_REGEX, PRODUCTIONS_REGEX
+
+        html = html.replace('\n', ' ').replace('\r', ' ')
+
+        columns = [m.group(1) for m in COLUMNS_REGEX.finditer(html)]
+        assert columns == EXPECTED_COLUMNS, 'Source format changed'
+
+        data = []
+        for time_m, prods_m in zip(TIMES_REGEX.finditer(html), PRODUCTIONS_REGEX.finditer(html)):
+            prods_list = prods_m.group(1, 2, 3, 4)
+            if 'null' in prods_list:
+                break
+            prods_list = [float(p) for p in prods_list]
+            time_str = time_m.group(1) + ' ' + time_m.group(2) + ':' + time_m.group(3)
+            self.append_datum(data, time_str, prods_list)
+
+        return data
+
+
+    def fetch_production(self, target_datetime: datetime.datetime) -> list:
+        if target_datetime is None:
+            url = 'https://tsoc.org.cy/total-daily-system-generation-on-the-transmission-system/'
+        else:
+            # convert target datetime to local datetime
+            url_date = arrow.get(target_datetime).to('Asia/Nicosia')
+            url = 'https://tsoc.org.cy/archive-total-daily-system-generation-on-the-transmission-system/?startdt={}&enddt=%2B1days'.format(
+                url_date.format('DD-MM-YYYY'))
+
+        res = self.session.get(url)
+        assert res.status_code == 200, 'CY parser: GET {} returned {}'.format(url, res.status_code)
+
+        data = self.parse_html(res.text)
+
+        if len(data) == 0:
+            self.logger.warning('No data returned')
```

Comment:
```suggestion
        if len(data) == 0:
            self.logger.warning(f'No production data returned for Cyprus on {url_date}', extra={'key': 'CY'})
```

Reply:
I removed `url_date` from the message as it's not set when fetching live data.